app/com/linkedin/drelephant/exceptions/core/ExceptionFingerprintingSpark.java

```diff
@@ -0,0 +1,368 @@
+      String inputLine;
+      String logLengthTrigger = LogLengthSchema.LOG_LENGTH.name().replace("_", " ");
+      try {
+        while ((inputLine = in.readLine()) != null) {
+          if (inputLine.toUpperCase().contains(logLengthTrigger)) {
+            logger.debug(" Processing of logs ");
+            driverLogProcessingForException(in, exceptions);
+            break;
+          }
+        }
+      } catch (IOException e) {
+        logger.error(" IO Exception while processing driver logs for ", e);
+      }
+      gracefullyCloseConnection(in, connection);
+    } catch (Exception e) {
+      logger.info(" Exception processing  driver logs ", e);
+      gracefullyCloseConnection(in, connection);
+    }
+    long endTime = System.nanoTime();
+    logger.info(" Total exception/error parsed so far - driver " + exceptions.size());
+    logger.info(" Time taken for driver logs " + (endTime - startTime) * 1.0 / (1000000000.0) + "s");
+  }
+
+  /**
+   * If unable to get log length , then read by default last 4096 bytes
+   * If log length is greater than threshold then read last some percentage of logs
+   * If log length is less than threshold then real complete logs .
+   * @param url
+   * @return
+   */
+  private String completeURLToQuery(String url) {
+    String completeURLToQuery = null;
+    BufferedReader in = null;
+    HttpURLConnection connection = null;
+    try {
+      URL amAddress = new URL(url);
+      connection = (HttpURLConnection) amAddress.openConnection();
+      connection.setConnectTimeout(JHS_TIME_OUT.getValue());
+      connection.setReadTimeout(JHS_TIME_OUT.getValue());
+      connection.connect();
+      in = new BufferedReader(new InputStreamReader(connection.getInputStream()));
+      String inputLine;
+      long logLength = 0l;
+      String logLengthTrigger = LogLengthSchema.LOG_LENGTH.name().replace("_", " ");
+      while ((inputLine = in.readLine()) != null) {
+        if (inputLine.toUpperCase().contains(logLengthTrigger)) {
+          logLength = getLogLength(inputLine.toUpperCase());
+          long startIndex = getStartIndexOfDriverLogs(logLength);
+          if (startIndex == 0) {
+            completeURLToQuery = url;
+          } else {
+            completeURLToQuery = url + "/stderr/?start=" + startIndex;
+          }
+          break;
+        }
+      }
+      gracefullyCloseConnection(in, connection);
+    } catch (Exception e) {
+      logger.error(" Exception while creating complete URL to query ", e);
+      gracefullyCloseConnection(in, connection);
+      return url;
+    }
+    return completeURLToQuery;
+  }
+
+  /**
+   *  Based on the log length , it will give the start index . Driver logs
+   *  will be read from given starting index
+   * @param logLength
+   * @return
+   * public for the purpose of testing . Visible for testing is not working
+   * if the unit test case is in scala.
+   */
+  public long getStartIndexOfDriverLogs(long logLength) {
```

Comment:
Can be a private method.

Reply:
It's for the purpose of testing . Since test cases are in scala visibleForTesting annotation is not working . Hence made it public